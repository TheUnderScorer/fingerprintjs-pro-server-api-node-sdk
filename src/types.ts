import type fetchFn from 'node-fetch';
import { components, paths } from './generatedApiTypes';

export enum Region {
  EU = 'EU',
  AP = 'AP',
  Global = 'Global',
}

export enum AuthenticationMode {
  AuthHeader = 'AuthHeader',
  QueryParameter = 'QueryParameter',
}

/**
 * Options for FingerprintJS server API client
 */
export interface Options {
  /**
   * Secret API key
   */
  apiKey: string;
  /**
   * Region of the FingerprintJS service server
   */
  region: Region;
  /**
   * Authentication mode
   * Optional, default value is AuthHeader
   */
  authenticationMode?: AuthenticationMode;

  /**
   * Optional fetch implementation
   * */
  fetch?: typeof fetchFn;
}

/**
 * More info: https://dev.fingerprintjs.com/docs/server-api#query-parameters
 */
export type VisitorHistoryFilter = paths['/visitors/{visitor_id}']['get']['parameters']['query'];

/**
 * More info: https://dev.fingerprintjs.com/docs/server-api#response
 */
export type VisitorsResponse =
  paths['/visitors/{visitor_id}']['get']['responses']['200']['content']['application/json'];
export type VisitorsError403 =
  paths['/visitors/{visitor_id}']['get']['responses']['403']['content']['application/json'] & {
    status: 403;
  };
export type VisitorsError429 =
  paths['/visitors/{visitor_id}']['get']['responses']['429']['content']['application/json'] & {
    status: 429;
    retryAfter: number;
  };

export type VisitorsError = VisitorsError403 | VisitorsError429;

export function isVisitorsError403(response: VisitorsError): response is VisitorsError403 {
  return response.status === 403;
}

export function isVisitorsError429(response: VisitorsError): response is VisitorsError429 {
  return response.status === 429;
}

export type EventResponse =
  paths['/events/{request_id}']['get']['responses']['200']['content']['application/json'];
export type EventError403 =
  paths['/events/{request_id}']['get']['responses']['403']['content']['application/json'];
export type EventError404 =
  paths['/events/{request_id}']['get']['responses']['404']['content']['application/json'];

type GenericEventError = EventError403 | EventError404;

type EventErrorCode<T extends GenericEventError> = T extends EventError403 ? 403 : 404;

export type EventError<T extends GenericEventError = GenericEventError> = T & {
  status: EventErrorCode<T>;
};

export function isEventError403(response: EventError): response is EventError<EventError403> {
  return response.status === 403;
}

export function isEventError404(response: EventError): response is EventError<EventError404> {
  return response.status === 404;
}

/**
 * More info: https://dev.fingerprintjs.com/docs/webhooks#identification-webhook-object-format
 */
export type VisitWebhook = components['schemas']['WebhookVisit'];
