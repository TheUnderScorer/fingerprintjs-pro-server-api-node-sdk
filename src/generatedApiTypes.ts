/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {
  '/events/{request_id}': {
    /** This endpoint allows you to get events with all the information from each activated product - BOTD and Fingerprinting. Use the requestId as a URL path :request_id parameter. This API method is scoped to a request, i.e. all returned information is by requestId. */
    get: operations['getEvent'];
  };
  '/visitors/{visitor_id}': {
    /** This endpoint allows you to get a history of visits with all available information. Use the visitorId as a URL path parameter. This API method is scoped to a visitor, i.e. all returned information is by visitorId. */
    get: operations['getVisits'];
  };
  '/webhook': {
    /** Fake path to describe webhook format. More information about webhooks can be found in the [documentation](https://dev.fingerprint.com/docs/webhooks) */
    trace: {};
  };
}

export interface components {
  schemas: {
    /**
     * PaginatedResponse
     * @description Fields `lastTimestamp` and `paginationKey` added when `limit` or `before` parameter provided and there is more data to show
     */
    Response: {
      visitorId: string;
      visits: {
        /**
         * @description Unique identifier of the user's identification request.
         * @example 1654815516083.OX6kx8
         */
        requestId: string;
        browserDetails: components['schemas']['BrowserDetails'];
        /** @description Flag if user used incognito session. */
        incognito: boolean;
        /**
         * Format: ipv4
         * @example 8.8.8.8
         */
        ip: string;
        ipLocation: components['schemas']['IPLocation'];
        /**
         * Format: int64
         * @description Timestamp of the event with millisecond precision in Unix time.
         * @example 1654815516086
         */
        timestamp: number;
        /**
         * Time
         * Format: date-time
         * @description Time expressed according to ISO 8601 in UTC format.
         * @example 2022-06-09T22:58:36Z
         */
        time: string;
        /**
         * Format: uri
         * @description Page URL from which identification request was sent.
         * @example https://some.website/path?query=params
         */
        url: string;
        /** @description A customer-provided value or an object that was sent with identification request. */
        tag: { [key: string]: unknown };
        /**
         * @description A customer-provided id that was sent with identification request.
         * @example someID
         */
        linkedId?: string;
        confidence: components['schemas']['Confidence'];
        /** @description Attribute represents if a visitor had been identified before. */
        visitorFound: boolean;
        /**
         * @example {
         *   "global": "2022-05-05T18:28:54.535Z",
         *   "subscription": "2022-06-09T22:58:05.576Z"
         * }
         */
        firstSeenAt: components['schemas']['StSeenAt'];
        /**
         * @example {
         *   "global": "2022-06-09T22:58:35.795Z",
         *   "subscription": null
         * }
         */
        lastSeenAt: components['schemas']['StSeenAt'];
      }[];
      /**
       * Format: int64
       * @description When more results are available (e.g., you scanned 200 results using `limit` parameter, but a total of 600 results are available), a special `lastTimestamp` top-level attribute is added to the response. If you want to paginate the results further in the past, you should use the value of this attribute.
       * @example 1654815517198
       */
      lastTimestamp?: number;
      /**
       * @description Visit's `requestId` of the last visit in the current page.
       * @example 1654815517198.azN4IZ
       */
      paginationKey?: string;
    };
    ErrorResponse: {
      /** ErrorResponseError */
      error?: {
        /** @example TokenRequired */
        code?: string;
        /** @example secret key is required */
        message?: string;
      };
    };
    ManyRequestsResponse: {
      /**
       * @description Error text.
       * @example request throttled
       */
      error: string;
    };
    WebhookVisit: {
      visitorId: string;
      clientReferrer?: string;
      /**
       * @description Unique identifier of the user's identification request.
       * @example 1654815516083.OX6kx8
       */
      requestId: string;
      browserDetails: components['schemas']['BrowserDetails'];
      /** @description Flag if user used incognito session. */
      incognito: boolean;
      /**
       * Format: ipv4
       * @example 8.8.8.8
       */
      ip: string;
      ipLocation: components['schemas']['IPLocation'];
      /**
       * Format: int64
       * @description Timestamp of the event with millisecond precision in Unix time.
       * @example 1654815516086
       */
      timestamp: number;
      /**
       * Time
       * Format: date-time
       * @description Time expressed according to ISO 8601 in UTC format.
       * @example 2022-06-09T22:58:36Z
       */
      time: string;
      /**
       * Format: uri
       * @description Page URL from which identification request was sent.
       * @example https://some.website/path?query=params
       */
      url: string;
      /** @description A customer-provided value or an object that was sent with identification request. */
      tag?: { [key: string]: unknown };
      /**
       * @description A customer-provided id that was sent with identification request.
       * @example someID
       */
      linkedId?: string;
      confidence: components['schemas']['Confidence'];
      /** @description Attribute represents if a visitor had been identified before. */
      visitorFound: boolean;
      /**
       * @example {
       *   "global": "2022-05-05T18:28:54.535Z",
       *   "subscription": "2022-06-09T22:58:05.576Z"
       * }
       */
      firstSeenAt: components['schemas']['StSeenAt'];
      /**
       * @example {
       *   "global": "2022-06-09T22:58:35.795Z",
       *   "subscription": null
       * }
       */
      lastSeenAt: components['schemas']['StSeenAt'];
    };
    /** Visit */
    Visit: {
      /**
       * @description Unique identifier of the user's identification request.
       * @example 1654815516083.OX6kx8
       */
      requestId: string;
      browserDetails: components['schemas']['BrowserDetails'];
      /** @description Flag if user used incognito session. */
      incognito: boolean;
      /**
       * Format: ipv4
       * @example 8.8.8.8
       */
      ip: string;
      ipLocation: components['schemas']['IPLocation'];
      /**
       * Format: int64
       * @description Timestamp of the event with millisecond precision in Unix time.
       * @example 1654815516086
       */
      timestamp: number;
      /**
       * Time
       * Format: date-time
       * @description Time expressed according to ISO 8601 in UTC format.
       * @example 2022-06-09T22:58:36Z
       */
      time: string;
      /**
       * Format: uri
       * @description Page URL from which identification request was sent.
       * @example https://some.website/path?query=params
       */
      url: string;
      /** @description A customer-provided value or an object that was sent with identification request. */
      tag?: { [key: string]: unknown };
      /**
       * @description A customer-provided id that was sent with identification request.
       * @example someID
       */
      linkedId?: string;
      confidence: components['schemas']['Confidence'];
      /** @description Attribute represents if a visitor had been identified before. */
      visitorFound: boolean;
      /**
       * @example {
       *   "global": "2022-05-05T18:28:54.535Z",
       *   "subscription": "2022-06-09T22:58:05.576Z"
       * }
       */
      firstSeenAt: components['schemas']['StSeenAt'];
      /**
       * @example {
       *   "global": "2022-06-09T22:58:35.795Z",
       *   "subscription": null
       * }
       */
      lastSeenAt: components['schemas']['StSeenAt'];
    };
    /** BrowserDetails */
    BrowserDetails: {
      /** @example Chrome */
      browserName: string;
      /** @example 101 */
      browserMajorVersion: string;
      /** @example 101.0.4951 */
      browserFullVersion: string;
      /** @example Windows */
      os: string;
      /** @example 10 */
      osVersion: string;
      /** @example Other */
      device: string;
      /** @example Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.41 Safari/537.36 */
      userAgent: string;
      botProbability?: number;
    };
    /** Confidence */
    Confidence: {
      /**
       * Format: float
       * @description The confidence score is a floating-point number between 0 and 1 that represents the probability of accurate identification.
       */
      score: number;
    };
    /** StSeenAt */
    StSeenAt: {
      /** Format: date-time */
      global: string | null;
      /** Format: date-time */
      subscription: string | null;
    };
    /** IPLocation */
    IPLocation: {
      /** @example 1000 */
      accuracyRadius: number;
      /**
       * Format: double
       * @example 37.75
       */
      latitude: number;
      /**
       * Format: double
       * @example -97.82
       */
      longitude: number;
      /** @example 130 00 */
      postalCode?: string;
      /**
       * Format: timezone
       * @example America/Chicago
       */
      timezone: string;
      /** IPLocationCity */
      city?: {
        /** @example Prague */
        name?: string;
      };
      /**
       * @example {
       *   "code": "US",
       *   "name": "United States"
       * }
       */
      country: components['schemas']['Location'];
      /**
       * @example {
       *   "code": "NA",
       *   "name": "North America"
       * }
       */
      continent: components['schemas']['Location'];
      subdivisions?: components['schemas']['Subdivision'][];
    };
    /** Location */
    Location: {
      /** @example US */
      code: string;
      /** @example United States */
      name: string;
    };
    Subdivision: {
      /** @example 10 */
      isoCode?: string;
      /** @example Hlavni mesto Praha */
      name?: string;
    };
    /** @description Contains all the information from each activated product - BOTD and Identification */
    ProductsResponse: {
      /** ProductsResponseIdentification */
      identification?: {
        /** ProductsResponseIdentificationData */
        data?: {
          /**
           * @description Unique identifier of the user's identification request.
           * @example 1654815516083.OX6kx8
           */
          requestId: string;
          browserDetails: components['schemas']['BrowserDetails'];
          /** @description Flag if user used incognito session. */
          incognito: boolean;
          /**
           * Format: ipv4
           * @example 8.8.8.8
           */
          ip: string;
          ipLocation: components['schemas']['IPLocation'];
          /**
           * Format: int64
           * @description Timestamp of the event with millisecond precision in Unix time.
           * @example 1654815516086
           */
          timestamp: number;
          /**
           * Time
           * Format: date-time
           * @description Time expressed according to ISO 8601 in UTC format.
           * @example 2022-06-09T22:58:36Z
           */
          time: string;
          /**
           * Format: uri
           * @description Page URL from which identification request was sent.
           * @example https://some.website/path?query=params
           */
          url: string;
          /** @description A customer-provided value or an object that was sent with identification request. */
          tag?: { [key: string]: unknown };
          /**
           * @description A customer-provided id that was sent with identification request.
           * @example someID
           */
          linkedId?: string;
          confidence: components['schemas']['Confidence'];
          /** @description Attribute represents if a visitor had been identified before. */
          visitorFound: boolean;
          /**
           * @example {
           *   "global": "2022-05-05T18:28:54.535Z",
           *   "subscription": "2022-06-09T22:58:05.576Z"
           * }
           */
          firstSeenAt: components['schemas']['StSeenAt'];
          /**
           * @example {
           *   "global": "2022-06-09T22:58:35.795Z",
           *   "subscription": null
           * }
           */
          lastSeenAt: components['schemas']['StSeenAt'];
          visitorId: string;
        };
      };
      /** ProductsResponseBotd */
      botd?: {
        data?: components['schemas']['BotdResult'];
      };
    };
    /** @description Contains event from activated products - BOTD and Identification */
    EventResponse: {
      products?: components['schemas']['ProductsResponse'];
    };
    /** @description Contains all the information from BOTD product */
    BotdResult: {
      /**
       * Format: ipv4
       * @description IP address of the requesting browser or bot.
       * @example 8.8.8.8
       */
      ip: string;
      /**
       * Time
       * Format: date-time
       * @description Time in UTC when the request from the JS agent was made. We recommend to treat requests that are older than 2 minutes as malicious. Otherwise, request replay attacks are possible
       * @example 2022-06-09T22:58:36Z
       */
      time: string;
      /**
       * Format: uri-reference
       * @description Page URL from which identification request was sent.
       * @example https://b.fpjs.sh/
       */
      url: string;
      /**
       * @example {
       *   "result": "notDetected"
       * }
       */
      bot: components['schemas']['BotdDetectionResult'];
    };
    /** @description Stores bot detection result */
    BotdDetectionResult: {
      /**
       * @description Bot detection result:
       *  * `notDetected` - the visitor is not a bot
       *  * `good` - good bot detected, such as Google bot, Baidu Spider, AlexaBot and so on
       *  * `bad` - bad bot detected, such as Selenium, Puppeteer, Playwright, headless browsers, and so on
       *
       * @enum {string}
       */
      result: 'notDetected' | 'good' | 'bad';
    };
  };
}

export interface operations {
  /** This endpoint allows you to get events with all the information from each activated product - BOTD and Fingerprinting. Use the requestId as a URL path :request_id parameter. This API method is scoped to a request, i.e. all returned information is by requestId. */
  getEvent: {
    parameters: {
      path: {
        /** Request ID */
        request_id: string;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          'application/json': components['schemas']['EventResponse'];
        };
      };
      /** Bad Request */
      400: {
        content: {
          'application/json': components['schemas']['ErrorResponse'];
        };
      };
      /** Forbidden */
      403: {
        content: {
          'application/json': components['schemas']['ErrorResponse'];
        };
      };
      /** Too Many Requests */
      429: {
        headers: {
          /** Indicates how long the user should wait before making a follow-up request. */
          'Retry-After'?: number;
        };
        content: {
          'application/json': components['schemas']['ManyRequestsResponse'];
        };
      };
    };
  };
  /** This endpoint allows you to get a history of visits with all available information. Use the visitorId as a URL path parameter. This API method is scoped to a visitor, i.e. all returned information is by visitorId. */
  getVisits: {
    parameters: {
      path: {
        visitor_id: string;
      };
      query: {
        /** Filter events by requestId */
        request_id?: string;
        /** Filter events by custom identifier */
        linked_id?: string;
        /** Limit scanned results */
        limit?: number;
        /** Used to paginate results */
        before?: number;
      };
    };
    responses: {
      /** OK */
      200: {
        content: {
          'application/json': components['schemas']['Response'];
        };
      };
      /** Too Many Requests */
      429: {
        headers: {
          /** Indicates how long the user should wait before making a follow-up request. */
          'Retry-After'?: number;
        };
        content: {
          'application/json': components['schemas']['ManyRequestsResponse'];
        };
      };
    };
  };
}

export interface external {}
