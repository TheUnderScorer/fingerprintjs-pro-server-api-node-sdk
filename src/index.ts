export * from './urlUtils';
export * from './serverApiClient';
export * from './types';
export * from './sealedResults';
