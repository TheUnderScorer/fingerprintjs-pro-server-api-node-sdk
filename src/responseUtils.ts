export function getJson<T>(response: Response) {
  return response.clone().json() as Promise<T>
}
