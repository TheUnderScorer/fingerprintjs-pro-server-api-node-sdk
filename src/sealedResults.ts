import { createDecipheriv } from 'crypto';
import { inflateRaw } from 'zlib';
import { promisify } from 'util';
import { EventResponse } from './types';
import { UnsealAggregateError, UnsealError } from './errors/unsealError';

const asyncInflateRaw = promisify(inflateRaw);

export enum DecryptionAlgorithm {
  Aes256Gcm = 'aes-256-gcm',
}

export interface DecryptionKey {
  key: Buffer;
  algorithm: DecryptionAlgorithm;
}

const SEALED_HEADER = Buffer.from([0x9e, 0x85, 0xdc, 0xed]);

export async function unsealEventsResponse(sealedData: Buffer, decryptionKeys: DecryptionKey[]) {
  const unsealed = await unseal(sealedData, decryptionKeys);
  const json = JSON.parse(unsealed) as EventResponse;

  if (!json.products) {
    throw new Error('Sealed data is not valid events response');
  }

  return json;
}

export async function unseal(sealedData: Buffer, decryptionKeys: DecryptionKey[]) {
  if (
    sealedData.subarray(0, SEALED_HEADER.length).toString('hex') !== SEALED_HEADER.toString('hex')
  ) {
    throw new Error('Invalid sealed data header');
  }

  const errors = new UnsealAggregateError([]);

  for (const decryptionKey of decryptionKeys) {
    switch (decryptionKey.algorithm) {
      case DecryptionAlgorithm.Aes256Gcm:
        try {
          return await unsealAes256Gcm(sealedData, decryptionKey.key);
        } catch (e) {
          errors.addError(new UnsealError(decryptionKey));
          continue;
        }

      default:
        throw new Error(`Unsupported decryption algorithm: ${decryptionKey.algorithm}`);
    }
  }

  throw errors;
}

async function unsealAes256Gcm(sealedData: Buffer, decryptionKey: Buffer) {
  const nonceLength = 12;
  const nonce = sealedData.subarray(SEALED_HEADER.length, SEALED_HEADER.length + nonceLength);

  const authTagLength = 16;
  const authTag = sealedData.subarray(-authTagLength);

  const ciphertext = sealedData.subarray(SEALED_HEADER.length + nonceLength, -authTagLength);

  const decipher = createDecipheriv('aes-256-gcm', decryptionKey, nonce).setAuthTag(authTag);
  const compressed = Buffer.concat([decipher.update(ciphertext), decipher.final()]);

  const payload = await asyncInflateRaw(compressed);

  return payload.toString();
}
