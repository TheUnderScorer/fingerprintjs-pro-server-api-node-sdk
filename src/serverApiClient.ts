import fetchFn from 'node-fetch';
import { getEventUrl, getVisitorsUrl } from './urlUtils';
import {
  VisitorHistoryFilter,
  VisitorsResponse,
  Region,
  Options,
  AuthenticationMode,
  EventResponse,
  EventError,
  isEventError403,
  isEventError404,
  VisitorsError,
  isVisitorsError403,
  isVisitorsError429,
} from './types';

export class FingerprintJsServerApiClient {
  public readonly region: Region;

  public readonly apiKey: string;

  public readonly authenticationMode: AuthenticationMode;

  protected readonly fetch: typeof fetchFn;

  /**
   * FingerprintJS server API client used to fetch data from FingerprintJS
   * @constructor
   * @param {Options} options - Options for FingerprintJS server API client
   */
  constructor(options: Readonly<Options>) {
    if (!options.region) {
      throw Error('Region is not set');
    }

    if (!options.apiKey) {
      throw Error('Api key is not set');
    }

    this.region = options.region;
    this.apiKey = options.apiKey;
    this.authenticationMode = options.authenticationMode ?? AuthenticationMode.AuthHeader; // Default auth mode is AuthHeader
    this.fetch = options.fetch ?? fetchFn;
  }

  public async getEvent(requestId: string) {
    if (!requestId) {
      throw new TypeError('requestId is not set');
    }

    const url =
      this.authenticationMode === AuthenticationMode.QueryParameter
        ? getEventUrl(requestId, this.region, this.apiKey)
        : getEventUrl(requestId, this.region);

    const headers = this.getHeaders();

    return this.fetch(url, {
      method: 'GET',
      headers,
    })
      .then(async (response) => {
        const jsonResponse = await response.json();
        if (response.status !== 200) {
          jsonResponse.status = response.status;
          throw jsonResponse as EventError;
        }
        return jsonResponse as EventResponse;
      })
      .catch((err) => {
        if (isEventError403(err) || isEventError404(err)) {
          throw err;
        }
        throw {
          status: 0,
          error: new Error(err.toString()),
        };
      });
  }

  /**
   * Gets history for the given visitor
   * @param {string} visitorId - Identifier of the visitor
   * @param {VisitorHistoryFilter} filter - Visitor history filter
   */
  public async getVisitorHistory(
    visitorId: string,
    filter?: VisitorHistoryFilter
  ): Promise<VisitorsResponse> {
    if (!visitorId) {
      throw TypeError('VisitorId is not set');
    }

    const url =
      this.authenticationMode === AuthenticationMode.QueryParameter
        ? getVisitorsUrl(this.region, visitorId, filter, this.apiKey)
        : getVisitorsUrl(this.region, visitorId, filter);
    const headers = this.getHeaders();

    return this.fetch(url, {
      method: 'GET',
      headers,
    })
      .then(async (response) => {
        const jsonResponse = await response.json();
        if (response.status === 200) {
          return jsonResponse as VisitorsResponse;
        }
        if (response.status === 429) {
          jsonResponse.retryAfter = response.headers.get('retry-after');
        }
        jsonResponse.status = response.status;
        throw jsonResponse as VisitorsError;
      })
      .catch((err) => {
        if (isVisitorsError403(err) || isVisitorsError429(err)) {
          throw err;
        }
        throw {
          status: 0,
          error: new Error(err.toString()),
        };
      });
  }

  private getHeaders() {
    return this.authenticationMode === AuthenticationMode.AuthHeader
      ? { 'Auth-API-Key': this.apiKey }
      : undefined;
  }
}
