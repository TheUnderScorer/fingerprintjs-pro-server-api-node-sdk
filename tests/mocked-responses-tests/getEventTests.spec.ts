import fetch from 'node-fetch';
import { Region } from '../../src/types';
import { FingerprintJsServerApiClient } from '../../src/serverApiClient';
import getEventResponse from './mocked-responses-data/external/get_event.json';

const { Response } = jest.requireActual('node-fetch');
jest.mock('node-fetch');

describe('[Mocked response] Get Event', () => {
  const apiKey = 'dummy_api_key';
  const existingRequestId = '1626550679751.cVc5Pm';

  const client = new FingerprintJsServerApiClient({ region: Region.EU, apiKey });

  test('with request_id', async () => {
    (fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(new Response(JSON.stringify(getEventResponse)))
    );

    const response = await client.getEvent(existingRequestId);

    expect(response).toMatchSnapshot();
  });
});
