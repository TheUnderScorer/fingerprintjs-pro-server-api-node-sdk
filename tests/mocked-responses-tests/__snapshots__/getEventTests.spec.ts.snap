// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Mocked response] Get Event with request_id 1`] = `
Object {
  "products": Object {
    "botd": Object {
      "data": Object {
        "bot": Object {
          "result": "notDetected",
        },
        "ip": "61.127.217.15",
        "time": "2019-05-21T16:40:13Z",
        "url": "https://www.example.com/login",
      },
    },
    "identification": Object {
      "data": Object {
        "browserDetails": Object {
          "browserFullVersion": "74.0.3729",
          "browserMajorVersion": "74",
          "browserName": "Chrome",
          "device": "Other",
          "os": "Windows",
          "osVersion": "7",
          "userAgent": "Mozilla/5.0 (Windows NT 6.1; Win64; x64) ....",
        },
        "confidence": Object {
          "score": 0.97,
        },
        "firstSeenAt": Object {
          "global": "2022-03-16T11:26:45.362Z",
          "subscription": "2022-03-16T11:31:01.101Z",
        },
        "incognito": true,
        "ip": "61.127.217.15",
        "ipLocation": Object {
          "accuracyRadius": 10,
          "city": Object {
            "name": "Dusseldorf",
          },
          "continent": Object {
            "code": "EU",
            "name": "Europe",
          },
          "country": Object {
            "code": "DE",
            "name": "Germany",
          },
          "latitude": 49.982,
          "longitude": 36.2566,
          "postalCode": "61202",
          "subdivisions": Array [
            Object {
              "isoCode": "63",
              "name": "North Rhine-Westphalia",
            },
          ],
          "timezone": "Europe/Dusseldorf",
        },
        "lastSeenAt": Object {
          "global": "2022-03-16T11:28:34.023Z",
          "subscription": null,
        },
        "linkedId": "somelinkedId",
        "requestId": "0KSh65EnVoB85JBmloQK",
        "time": "2019-05-21T16:40:13Z",
        "timestamp": 1582299576512,
        "url": "https://www.example.com/login",
        "visitorFound": true,
        "visitorId": "Ibk1527CUFmcnjLwIs4A9",
      },
    },
  },
}
`;
